File: C++/MPI_boilerplate.cpp
/**
 * MPI Boilerplate for C++
 * @Author  : Otto Lindblom
 * @Link    : https://github.com/prOttonicFusion/MPI-boilerplates
*/

#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>

using namespace std;

int main(int argc, char **argv)
{
  const int tag = 50;
  int numProc, rank, nameLenght;
  char procName[MPI_MAX_PROCESSOR_NAME];
  MPI_Status status;

  // Initialize MPI
  if (MPI_Init(&argc, &argv) != MPI_SUCCESS)
  {
    cout << "MPI initialization failed" << endl;
    return (-1);
  }

  // Query process information
  MPI_Comm_size(MPI_COMM_WORLD, &numProc);       // numProc = number of processes/CPUs
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);          // rank = id of current process
  MPI_Get_processor_name(procName, &nameLenght); // procName = name of current process


  /**
   * 
   * * MPI code goes here * *
   * 
   * Typical commands include: 
   * 
   * Send/receive data:
   * MPI_Send(void* data, int count, MPI_Datatype datatype, int destination, int tag, MPI_Comm communicator)
   * MPI_Recv(void* data, int count, MPI_Datatype datatype, int source, int tag, MPI_Comm communicator, MPI_Status* status)
   * 
   * Reduce data from all processes to one (e.g. global sum when using op = MPI_SUM):
   * MPI_Reduce(void* send_data, void* recv_data, int count, MPI_Datatype datatype, MPI_Op op, int rootProcID, MPI_Comm communicator)
   * 
   * Send data to all processes:
   * MPI_Bcast(void* data, int count, MPI_Datatype datatype, int rootProcID, MPI_Comm communicator)
   * 
  */


  // Finalize MPI & quit program
  MPI_Finalize();
  return (0);
}

File: Fortran/MPI_boilerplate.f90
!
! MPI Boilerplate for Fortran
! @Author  : Otto Lindblom
! @Link    : https://github.com/prOttonicFusion/MPI-boilerplates
!

program mpi_boilerplate

use mpi
implicit none

  integer, parameter :: tag = 50
  integer :: numproc, rank, namelength, ierr
  character(len=22) :: procname = "MPI_MAX_PROCESSOR_NAME"

  ! Initialize MPI
  call mpi_init(ierr)
  if (ierr /= 0) then 
    print *, "MPI initialization failed"
  end if

  ! Query process information
  call mpi_comm_size(MPI_COMM_WORLD, numproc, ierr);       ! numproc = number of processes/CPUs
  call mpi_comm_rank(MPI_COMM_WORLD, rank, ierr);          ! rank = id of current process
  call mpi_get_processor_name(procname, namelength, ierr); ! procname = name of current process

  ! 
  ! * MPI code goes here * *
  ! 
  ! Typical commands include: 
  ! 
  ! Send/receive data:
  ! call mpi_send(void* data, int count, MPI_Datatype datatype, int destination, int tag, MPI_Comm communicator, ierr)
  ! call mpi_recv(void* data, int count, MPI_Datatype datatype, int source, int tag, MPI_Comm communicator, MPI_Status* status, ierr)
  ! 
  ! Reduce data from all processes to one (e.g. global sum when using op = MPI_SUM):
  ! call mpi_reduce(void* send_data, void* recv_data, int count, MPI_Datatype datatype, MPI_Op op, int rootProcID, MPI_Comm communicator, ierr)
  ! 
  ! Send data to all processes:
  ! call mpi_bcast(void* data, int count, MPI_Datatype datatype, int rootProcID, MPI_Comm communicator, ierr)
  ! 

  ! Finalize MPI & quit program
  call mpi_Finalize(ierr)

end program mpi_boilerplate
